data/moves.js

```diff
@@ -13038,8 +13038,13 @@ exports.BattleMovedex = {
 		flags: {protect: 1, authentic: 1, mystery: 1},
 		onHit: function (target, source) {
 			if (source.template && (source.template.num === 493 || source.template.num === 773)) return false;
+			if (target.getTypes() === ['???']) return false;
```

Comment:
... Do you ever test your code?

Reply:
The GitHub client is absolute trash so the only way for me to test code is to actually write tests, and I didn't think this was something that needed to be tested every time someone wanted to fix a chat command, but if you insist, I'll spend 2 hours trying to make the tests work only to find out that Burn Up has undocumented behavior (probably a bug) that was making the tests fail for no reason, and that Marty's description of how Reflect Type is supposed to work is at best vague and at worst inaccurate.

Reply:
"Don't get mad at me because it's inconvenient for you to test your own code." I'm not mad at you, just grumpy.

Reply:
What's so vague or inaccurate about it? If you have a secondary type, ignore typeless; if you have an added type, typeless becomes Normal; typeless alone fails. Could not be simpler. I could write it in excruciating detail and make it twice as long, but why?

Reply:
Then shouldn't this be Normal/Flying/Ghost? http://www.smogon.com/forums/threads/pokemon-sun-moon-battle-mechanics-research.3586701/page-72#post-7508472

Reply:
See the first point: "If you have a secondary type, ignore typeless".
Do I really have to say "If you have a secondary type and an added type, also ignore typeless"?

Reply:
This is exactly the sort of thing you should write tests for; tests are also good documentation.

Reply:
That is not what the description in the dex says. "If the target's current types include typeless and a non-added type, typeless is ignored. If the target's current types include typeless and an added type from Forest's Curse or Trick-or-Treat, typeless is copied as the Normal type instead." To me, this says "If you have no added type, ignore typeless. If you have an added type, copy typeless as normal." What you're saying now is "If you have *a secondary type,* ignore typeless. If you have *no secondary type,* but do have an added type, copy typeless as Normal." The dex description doesn't even mention secondary types.

Reply:
At least, it does the first time, but it's not clear whether the second time is an exception or an alternative.

Reply:
And speaking of documentation, is Burn Up supposed to remove the user's added type?

Reply:
To be fair, Marty, your description is hard for me to understand, too. It's an annoyingly complex mechanic and it'll probably be hard to phrase well.

Current phrasing:

> Causes the user's types to become the same as the current types of the target. If the target's current types include typeless and a non-added type, typeless is ignored. If the target's current types include typeless and an added type from Forest's Curse or Trick-or-Treat, typeless is copied as the Normal type instead. Fails if the user is an Arceus or a Silvally, or if the target's current type is typeless alone.

I might phrase it more like:

> Causes the user's types to become the same as the current types of the target. Fails if the user is an Arceus or a Silvally. Typeless is not copied: If this would leave the user with no type at all, fails. If this would leave the user with no type other than an added type (from Forest's Curse or Trick-or-Treat), the user's primary type becomes Normal.

Reply:
You could easily clear up that ambiguity with one word: "If the target's current type includes *only* typeless and an added type from Forest's Curse or Trick-or-Treat, typeless is copied as the Normal type instead."

Reply:
Yeah the second "includes" should just be "are" then. I'll mull over your suggestions and come up with something.